corehq/apps/locations/fixtures.py

```diff
@@ -207,14 +216,428 @@ def should_sync_flat_fixture(project):
 )
 
 
+int_field = IntegerField()
+int_array = ArrayField(int_field)
+
+
+class Array(Func):
+    function = "Array"
+    template = '%(function)s[%(expressions)s]'
+    output_field = int_array
+
+
+class array_append(Func):
+    function = "array_append"
+    output_field = int_array
+
+
+@Field.register_lookup
+class EqualsAny(Lookup):
+    # Can't use Func for ANY(...) because functions are wrapped in parens
+    # resulting in SQL like `... = (ANY(...))` which is not valid.
+    lookup_name = "eq_any"
+
+    def as_postgresql(self, compiler, connection):
+        lhs, lhs_params = self.process_lhs(compiler, connection)
+        rhs, rhs_params = self.process_rhs(compiler, connection)
+        params = lhs_params + rhs_params
+        return '%s = ANY(%s)' % (lhs, rhs), params
+
+
+class WrapSQL(RawSQL):
+    """Wrap ORM-generated SQL expression(s) with raw SQL
+
+    Desparate measures for desparate times. Any `{named_sub_expression}`
+    slugs in the "wrapper" SQL string will be filled with the
+    corresponding ORM SQL expression passed as a keyword argument.
+    """
+
+    def __init__(self, _sql, **expressions):
+        self.sql = _sql
+        output_field = expressions.pop("output_field", Field())
+        self.expressions = expressions
+        # purposely skip RawSQL.__init__ by calling its super
+        super(RawSQL, self).__init__(output_field=output_field)
+
+    def __repr__(self):
+        return "{}({}, {})".format(
+            self.__class__.__name__, self.sql, self.expressions)
+
+    def as_sql(self, compiler, connection):
+        expressions = {}
+        params = []
+        for key, expr in self.expressions.items():
+            if hasattr(expr, "get_compiler"):
+                comp = expr.get_compiler(connection=connection)
+                expr_sql, expr_params = comp.as_sql()
+            else:
+                expr_sql, expr_params = expr.as_sql(compiler, connection)
+            expressions[key] = expr_sql
+            params.extend(expr_params)
+        return self.sql.format(**expressions), params
+
+    def get_group_by_cols(self):
+        return [self]
+
+
 def get_location_fixture_queryset(user):
     if toggles.SYNC_ALL_LOCATIONS.enabled(user.domain):
         return SQLLocation.active_objects.filter(domain=user.domain).prefetch_related('location_type')
 
     timing = TimingContext("get_location_fixture_queryset")
     with timing("mptt"):
         mptt_set = mptt_get_location_fixture_queryset(user)
-    return ComparedQuerySet(mptt_set, timing)
+    with timing("cte"):
+        cte_set = cte_get_location_fixture_queryset(user)
+    return ComparedQuerySet(mptt_set, cte_set, timing)
+
+
+def cte_get_location_fixture_queryset(user):
+
+    # There may be ambiguities in location type configurations that could
+    # cause undefined outcomes:
```

Comment:
what does this mean in terms of rolling this out, do those location type configurations need to be checked to see if they exist currently, or restricted from being created in the ui?

Reply:
These are general observations I made while converting the existing Python logic (which executes several SQL queries) into a single SQL query. None of them are new with this PR.